### DISCUSSION TITLE
How do I employ cache busting in a template?

### DISCUSSION THREAD

**Discussion by ugyballoons** [aio-libs/aiohttp-jinja2#590] [Q&A]
Hi,
I'm enjoying working with aiohttp and jinja2.
It's not clear to me how to employ cache busting. Please could someone help?
In the setup I have:
```
sub_app.add_routes(
    [
        web.static(
            "/static",
            Path(__file__).parent / "static",
            name="static",
            append_version=True,
        ),
    ]
)
sub_app["static_root_url"] = "/path-to/static"
env = aiohttp_jinja2.get_env(sub_app)
env.globals.update(
    zip=zip,
    url_for=web.Resource.url_for
)
```
and in the template I have:
`<link rel="stylesheet" href="{{ url_for('static', filename='stylesheets/main.css') }}">`

I feel that I shouldn't be passing `url_for` in the globals but if I don't, then jinja complains that there's no such function. Where am I going wrong?

**Comment by Dreamsorcerer** [comment]
I think your question is in the wrong repo, but if you can't access `url_for()`, then you've probably missed the setup() call, as shown in the usage:
https://aiohttp-jinja2.readthedocs.io/en/stable/#usage

I think each sub app will need to be setup separately from the root app (maybe that should be mentioned in the docs).

**Reply by ugyballoons** [reply]
Thanks for the response!
I could have been clearer- I do have:
```
    sub_app = web.Application()
    aiohttp_jinja2.setup(
        sub_app,
        loader=jinja2.FileSystemLoader(Path(__file__).parent / "templates"),
    )
```
before the lines quoted at the top. The sub-app is running & responding to requests.
From what you're saying, can I take it that `url_for()` ought to be accessible in the templates?
If I don't pass it in explicitly, and I suspect wrongly, I get:
`jinja2.exceptions.UndefinedError: 'url_for' is undefined`

**Reply by Dreamsorcerer** [reply]
Sorry, I didn't even think to check you were using the correct function name...
The functions available in templates are `url()` and `static()`: https://aiohttp-jinja2.readthedocs.io/en/stable/#default-globals

**Reply by ugyballoons** [reply]
I can use `static('stylesheets/main.css')` in the template.
I hoped that the flag `append_version=True` would then see that a hash was applied to the end of the resulting url in the html but it's not the case. As far as I can make out from the source, only `url_for()` applies the version hash of the resource. It would be great if `static()` would do that...

**Reply by Dreamsorcerer** [reply]
Well, the append_version is on the route, so it seems like something that _should_ work. Feel free to look into the code and try to come up with a solution.

static() is at: https://github.com/aio-libs/aiohttp-jinja2/blob/master/aiohttp_jinja2/helpers.py#L61
It seems to currently just get `app["static_root_url"]`, which is just manually entered by the user. This has always seemed slightly awkward to me, but I guess the main purpose is to point to another location for static sources (e.g. a CDN or node dev server). Maybe we need to add append_version logic into this function...

I think to use the existing append_version feature, you would need to use `.url_for()` on the static router.
Your previous code just assigns url_for to a random unbound method, which can't be right. The normal way in aiohttp, is using the method on the router, accessed through your app: https://docs.aiohttp.org/en/stable/web_quickstart.html#reverse-url-constructing-using-named-resources

**Reply by ugyballoons** [reply]
That last link pointed me to the solution, thank you!

I hadn't put twos together but it's there in the signature for 
```
aiohttp.web.static(prefix, path, *, name=None, expect_handler=None, chunk_size=256 * 1024, show_index=False, follow_symlinks=False, append_version=False)
```
That *is* a named resource it's setting as a route, so all I need to do in the template is
```
 <link rel="stylesheet" href="{{ url('static', filename='stylesheets/main.css') }}">
```
and it works.

As regards to the warnings that are in the aiohttp docs for only using the app to serve files in development, I may have got the wrong end of the stick but isn't cache busting by appending a hash query to URLs effective, however those files are being served? I mean, if a reverse proxy is being used, isn't that the same difference to a browser, where the request to ignore the cache is actually being done?

**Reply by Dreamsorcerer** [reply]
> and it works.

Great, maybe we can add something to the docs about this. Although it is unclear to me how this should work in relation to `static()` currently.

> As regards to the warnings that are in the aiohttp docs for only using the app to serve files in development, I may have got the wrong end of the stick but isn't cache busting by appending a hash query to URLs effective, however those files are being served?

You'd have to point me to which part you are reading. Static files should only be getting served from the app during development, but if you have nginx or something to serve the files, then the app is not serving them, even if you leave the code in so it is capable of it. It just avoids running expensive Python processes to serve a simple file.

**Reply by ugyballoons** [reply]
> Great, maybe we can add something to the docs about this. Although it is unclear to me how this should work in relation to static() currently.

Yes, it would be useful, I think, to demonstrate how `url()` hooks up with static routing. The named `filename` attribute isn't mentioned at all. Is this something I could help with?

> Static files should only be getting served from the app during development, but if you have nginx or something to serve the files, then the app is not serving them, even if you leave the code in so it is capable of it. It just avoids running expensive Python processes to serve a simple file.

Good to know!

**Reply by Dreamsorcerer** [reply]
> Yes, it would be useful, I think, to demonstrate how `url()` hooks up with static routing. The named `filename` attribute isn't mentioned at all. Is this something I could help with?

Feel free to make a PR if you have a suggestion. Otherwise, I'll try and take a look some time later.
